File: Project/Simulator/Abilities/Ability.java
package Simulator.Abilities;

import Simulator.Organisms.Organism;

public abstract class Ability {
    private final int defaultCooldown;
    private final int defaultDuration;
    private int cooldown = 0;
    private int duration = 0;
    private boolean toggle = false;

    public Ability(int cooldown, int duration) {
        this.defaultCooldown = cooldown;
        this.defaultDuration = duration + 1;
    }

    public Ability(int cooldown, int duration, int defaultCooldown, int defaultDuration) {
        this.cooldown = cooldown;
        this.duration = duration;
        this.defaultCooldown = defaultCooldown;
        this.defaultDuration = defaultDuration;
    }

    public boolean isReady() {
        return cooldown == 0;
    }

    public boolean isActive() {
        return duration > 0;
    }

    public int getCooldown() {
        return cooldown;
    }

    public int getDuration() {
        return duration;
    }

    public boolean checkToggle() {
        return toggle;
    }

    public void flipToggle() {
        toggle = !toggle;
    }

    public void use() {
        if (isReady() && !isActive()) {
            cooldown = defaultCooldown;
            duration = defaultDuration;
        }
    }

    public void update() {
        if (duration > 0) {
            duration--;
        } else if (cooldown > 0) {
            cooldown--;
        }
    }

    public abstract void effect(Organism user, Organism other);
}

File: Project/Simulator/Abilities/Ability.py
from abc import ABC, abstractmethod
from Simulator.Organisms.Organism import Organism
from typing import Optional


class Ability(ABC):
    def __init__(self, cooldown, duration, default_cooldown=-1, default_duration=-1):
        if (default_cooldown == -1) or (default_duration == -1):
            self.___default_cooldown = cooldown
            self.___default_duration = duration + 1
            self.__cooldown = 0
            self.__duration = 0
        else:
            self.___default_cooldown = default_cooldown
            self.___default_duration = default_duration
            self.__cooldown = cooldown
            self.__duration = duration
        self.__toggle = False

    def isReady(self):
        return self.__cooldown == 0

    def isActive(self):
        return self.__duration > 0

    def getCooldown(self):
        return self.__cooldown

    def getDuration(self):
        return self.__duration

    def checkToggle(self):
        return self.__toggle

    def flipToggle(self):
        self.__toggle = not self.__toggle

    def use(self):
        if (self.isReady() and not self.isActive()):
            self.__cooldown = self.___default_cooldown
            self.__duration = self.___default_duration

    def update(self):
        if self.__duration > 0:
            self.__duration -= 1
        elif self.__cooldown > 0:
            self.__cooldown -= 1

    @abstractmethod
    def effect(self, user: Organism, other: Optional[Organism] = None):
        pass
